File: arduino/uvgad/Pulse32.cpp
#include <stddef.h>
#include <stdint.h>
#include <string.h>

#include "CRC.h"
#include "ErrorCode.h"
#include "Pulse32.h"

int Pulse32::serialize(uint8_t *data, size_t &len) const
{
  if(!data || len < PKT_SIZE)
  {
    return(ErrorCode::ERR_PARAMS);
  }
  
  memcpy(data, reinterpret_cast<const uint8_t *>(&m_pkt), PKT_SIZE);
  len = PKT_SIZE;
  
  return(ErrorCode::NO_ERROR);
}

int Pulse32::parse(const uint8_t *data)
{
  if(!data)
  {
    return(ErrorCode::ERR_PARAMS);
  }
  
  memcpy(reinterpret_cast<uint8_t *>(&m_pkt), data, PKT_SIZE);
  return(ErrorCode::NO_ERROR);
}

void Pulse32::setField(unsigned int id, int32_t value)
{
  if(id > PKT_MAX_PAYLOAD_ENTRIES)
  {
    id = 0;
  }
  
  m_pkt.flags |= 0x8000 >> id;
  m_pkt.payload[id] = value;
}

int32_t Pulse32::getField(unsigned int id) const
{
  if(id > PKT_MAX_PAYLOAD_ENTRIES)
  {
    id = 0;
  }
  
  return(m_pkt.payload[id]);
}

void Pulse32::setFlags(uint16_t flags)
{
  m_pkt.flags = flags;
}

int16_t Pulse32::getFlags() const
{
  return(m_pkt.flags);
}

bool Pulse32::isFieldSet(unsigned int id) const
{
  if(id > PKT_MAX_PAYLOAD_ENTRIES)
  {
    id = 0;
  }
  
  return((m_pkt.flags & (0x8000 >> id)) ? true : false);
}

void Pulse32::clear()
{
  m_pkt.flags = 0;
  memset(m_pkt.payload, 0, sizeof(m_pkt.payload));
}

File: SerialUsbDriver/src/edu/repo/ucla/serialusbdriver/Pulse32.java
package edu.repo.ucla.serialusbdriver;


import java.nio.ByteBuffer;
import java.nio.ByteOrder;

public class Pulse32
{
	public static final int PKT_SIZE = 32;
	public static final int PKT_MAX_PAYLOAD_ENTRIES = 7;
	
	protected class Pulse32Pkt
	{
		protected short flags;
		protected final short padding;
		protected int payload[] = new int[PKT_MAX_PAYLOAD_ENTRIES];
		
		Pulse32Pkt()
		{
			flags = 0;
			padding = 0;
			
			for(int cursor = 0; cursor < payload.length; cursor++)
			{
				payload[cursor] = 0;
			}
		}
	}
	
	protected final Pulse32Pkt mPkt;
	
	public Pulse32()
	{
		mPkt = new Pulse32Pkt();
	}
	
	byte[] serialize()
	{
		ByteBuffer buffer = ByteBuffer.allocate(PKT_SIZE);
		buffer.order(ByteOrder.LITTLE_ENDIAN);
		
		buffer.putShort(mPkt.flags);
		buffer.putShort(mPkt.padding);
		
		for(int cursor = 0; cursor < PKT_MAX_PAYLOAD_ENTRIES; cursor++)
		{
			buffer.putInt(mPkt.payload[cursor]);
		}
		
		return(buffer.array());
	}
	
	int parse(final byte data[])
	{
		if(data == null)
		{
			return(ErrorCode.ERR_PARAMS);
		}
		
		ByteBuffer buffer = ByteBuffer.wrap(data);
		buffer.order(ByteOrder.LITTLE_ENDIAN);
		
		mPkt.flags = buffer.getShort();
		buffer.getShort();			// Increments the buffer pointer to ignore padding
		
		for(int cursor = 0; cursor < PKT_MAX_PAYLOAD_ENTRIES; cursor++)
		{
			mPkt.payload[cursor] = buffer.getInt();
		}
		
		return(PKT_SIZE);
	}

	void setField(int id, int value)
	{
		if(id < 0 || id > PKT_MAX_PAYLOAD_ENTRIES)
		{
			id = 0;
		}
		
		mPkt.flags |= (short) (0x8000 >>> id);
		mPkt.payload[id] = value;
	}
	
	int getField(int id)
	{
		if(id < 0 || id > PKT_MAX_PAYLOAD_ENTRIES)
		{
			id = 0;
		}
		
		return(mPkt.payload[id]);
	}
	
	void setFlags(short flags)
	{
		mPkt.flags = flags;
	}
	
	short getFlags()
	{
		return(mPkt.flags);
	}
	
	boolean isFieldSet(int id)
	{
		if(id < 0 || id > PKT_MAX_PAYLOAD_ENTRIES)
		{
			id = 0;
		}
		
		return((mPkt.flags & (short) (0x8000 >>> id)) != 0 ? true : false);
	}
	
	void clear()
	{
		mPkt.flags = 0;
		
		for(int cursor = 0; cursor < PKT_MAX_PAYLOAD_ENTRIES; cursor++)
		{
			mPkt.payload[cursor] = 0;
		}
	}
}
